### DISCUSSION TITLE
How to center certain windows by default and remember position and size?

### DISCUSSION THREAD

**Discussion by pentago** [rxhanson/RectanglePro-Community#449] [General]
When purchasing, I thought this would solve the problem I have but wasn't be able to configure it to behave like this.
Does anyone have an idea how to make this happen?
Thanks in advance!

**Comment by rxhanson** [comment]
Thanks for purchasing. This sounds like something you should be able to accomplish with App Layouts. Can you provide more details about what exactly you're looking to do?

**Comment by pentago** [comment]
I’ re tried app layouts and in 3-monotor setup, it doesnt really work the way I want. I mean widnows stay at alproximate places where I save them but don’t remember exact positions.

Thats’s the reason why I asked the question here, can it somehow remember exact positions of windows and sizes of windows and set this in stone for eternity?

Could the problem be that my laptop and external displays do not have the same resolution?
________________________________

**Reply by pentago** [reply]
@rxhanson ping

Still having this issue. Somehow when I connect my 2 external displays to my macbook, the window positioning and sizing is not the way I left them.

I think window size/position saving is broken or perhaps never did work.
From what I can see, when I arrange windows the way I want them and save layout, after turning external screens off and on, all windows get packed on my main screen and do not restore back to their saved layout positions

Perhaps I don't understand how exactly RectanglePro works and what it does but from reading settings descriptions, it's logical to make the assumption that the term "saved layout" means the windows will end up in their saved positions when external screens are connected, right?

If I'm misunderstanding how this thing works, I'd like to be educated, please :)

Most of the time I'm using 7 windows (1 on main, 3 on left and 3 on right external screen) and I hate turning those apps and arranging their windows each time, I'd appreciate if Rectangle can do this for me.

Somehow, this doesn't do the trick for me:
![image](https://github.com/rxhanson/RectanglePro-Community/assets/876756/16d0d188-eb97-4fba-b3d6-0fe723bfd2a6)

Basically, no matter what config combo I do, I never end up achieving what I describe.

**Reply by rxhanson** [reply]
Sorry for not responding here. Rectangle Pro is simply applying layouts whenever it receives the notification from macOS that a display has been added (or removed, if configured). It's possible that there could be an issue specific to the displays you have where the layout is applied sooner than the display is ready, and the layout doesn't actually take effect. If you test out the layouts you made by assigning them keyboard shortcuts, do the windows go where you expect them to go? Maybe all we need is an adjustable delay for when to apply the layout after the display is connected?

Resolution shouldn't be an issue here, unless you have moved around the displays in System Settings and the absolute positions saved off no longer apply (probably not the case).

Also note that in the most recent version of Rectangle Pro, I put in a fix so that windows of the same app will be differentiated, if possible, and placed in the position you had initially saved. Prior to this, the windows would be placed down in order of z-index. You would have to re-create the saved layouts to take advantage of this, unfortunately.

**Reply by pentago** [reply]
You might be right on delay. I noticed when I got these two screens that from the moment I plug the USB-C cable to laptop, it takes couple of seconds for image to appear on them both so perhaps having ability to adjust amount of time before applying layout might be the solution.

Or perhaps, if Rectangle could poll on schedule to check if there are screens that were previously present and reattach windows to them once they appear could also work but I guess at expense on some performance.

I don't know for how this could work but from reading MacOS-related stuff on window position saving/restoring I got feel that OS is extremely finicky about this

**Reply by rxhanson** [reply]
In the next release, I'll start by adding in some terminal command configuration for delay. Going with some sort of polling introduces a few scenarios that aren't ideal, but at least getting a ballpark number on how long a delay needs to be gives a good starting point for getting a better solution in place.

**Reply by pentago** [reply]
I just measured how much time it takes screens to display stuff from the moment I plug them in, it's 10 seconds.

**Reply by rxhanson** [reply]
Thanks! I'll post back when I have a new release that adds the delay configuration.

**Reply by rxhanson** [reply]
Ok, in the latest version, 3.0.20, I have added a configurable delay. Here is an example terminal command for it:

```
defaults write com.knollsoft.Hookshot displayConnectedDelay -float 10
```

The value being set here is 10 (seconds). You'll have to restart the app afterward, if it's running. So you could run that command, then update the app and see how it goes. 

It would be nice if there were some sort of "display ready for window adjustments" notification that we could get from macOS, but as far as I know there is nothing of the sort. Allowing a configurable delay might be as good as it gets here :(
